vlm-tool.1
.TH VLM-TOOL "1" "Sep 2013" "MegaCoder.com" "User Commands"
.SH NAME
vlm-tool \- intelligent grep(1) for /var/log/messages files
.SH SYNOPSIS
.B vlm-tool
[\fIOPTION\fR]... [\fIfile\fR...]
.SH DESCRIPTION
.PP
This man(1) page gives only the command-line usage.
I know these are a lot of options; see the \fIREADME\fP file for full details.
.TP
\fB\-X\fP
increase level of debug output.
Not useful to anyone other than the developer and probably not even him.
.TP
\fB\-a\fP \fIpattern\fP
Add \fIpattern\fP to list of regular expressions to find.
.TP
\fB\-A\fP \fIfile\fP
Add list of regular expressions from \fIfile\fP, one pattern per line.
.TP
\fB\-c\fP
Colorize the output.
This assigns each host a different color, so that finding related entries is easier.
You do not get any choice about the color selection.
.TP
\fB\-d\fP \fIregion\fP
Change the date parsing to \fIregion\fP.
The default is \fIUSA\fP.
Region \fIISO\fP attempts to decode using the ISO 8601 standard and most of its variations.
.TP
\fB\-G\fP \fIsecs\fP
Recognize a time gap if two log entries for this host are \fIsecs\fP apart.
.PP
Gaps are calculated after uninteresting lines have be dropped.
The intent is to prevent flooding log entries from obscuring the intervals between interesting lines.
.TP
\fB\-g\fP
Flag gaps in the log files.
A gap is recognized when the timestamps on two consecutive line from the same host are more than \fIN\fP seconds apart.
By default, \fIN\fP is 10 minutes.
.TP
\fB\-I\fP \fIfile\fP
Read a list of patterns, one per line, that identifies log entries to be ignored.
.TP
\fB\-i\fP \fIpattern\fP
Ignore log entries matching \fIpattern\fP.
.TP
\fB\-l\fP
Output the table of default trigger patterns.
No other switches are honored if \fB-l\fP is used.
.TP
\fB\-m\fP
Output all log entries.
Entries matching a trigger pattern will be marked with the \fIthumb\P.
.PP
Some log entries consist of multiple lines and form a \fIstanza\fP.
For example, the string \fBCall Trace:\fP marks a register dump and stack trace.
.TP
\fB\-n\fP
Do not load the table of built-in trigger rules.
.TP
\fB\-N\fP
Prefix each output line with a number.
The first line output is one, the next two, and so on.
Ignored or unmatched lines are not counted.
.TP
\fB\-o\fP \fIofile\fP
Write the output to \fIofile\fP instead of the default \fIstdout\fP.
.TP
\fB\-r\fP
Show the trigger rule which caused this line to be selected.
.TP
\fB\-s\fP
Output some statistics at the end.
No really sure why I did this.
.TP
\fB\-t\fP \fIthumb\fP
Use \fIthumb\fP to mark a matched log entry.
The default thumb is \fI- \fP, to make searching for an interesting log entry simple but you may prefer a different pattern.
Any size thumb will work.
.TP
\fB\-w\fP \fIwhen\fP
This is the approximate time the issue occurred.
The format is exactly the same as the \f(CW/var/log/messages\fP entries.
.TP
\fB\-W\fP \fIwindow\fP
A fuzzy time delta to capture entries before and after the incident time set by the \fB-w\fP switch.
By default, the delta is 24 hours.
The \fIwindow\fP is the number of seconds and can be followed by \fBs\fP for seconds (default), \fBh\fP for hours, \fBd\fP for days, \fBw\fP for weeks.
.TP
\fB\-v\fP
Print the program version and exit.
.TP
\fB\-y\fP \fIyear\fP
Normally log entries do not include the year value.
The current year is assumed, but if you can specify a different year if needed.

.PP
Visually scanning a \fI/var/log/messages\fP file with hundreds of thousands of entries is boring.
On a single-server setup, this tool identifies interesting log file entries for you.
.PP
In a clustered environment, the input files are sorted internally by host and then by date.
.SH AUTHOR
Written by Tommy Reynolds <Tommy.Reynolds@MegaCoder.com>
.SH COPYRIGHT
Copyright \(co 2013 Tommy Reynolds; all rights reserved; international copyright secured.
.PP
This code is licensed under the GPL v2, and no other.
\fBvlm-tool\fP detects several typs of the stanzas and tries to mark the entire set of lines.
.SH "SEE ALSO"
.PP
\fIlogger(\fP1\fI)\fP
\fIgrep(\fP1\fI)\fP
\fIsyslog(\fP3\fI)\fP


ChangeLog
* Fri Jul 27 2018 Tommy Reynolds
- (fa8edf6) Highlight hypervisor detection. (HEAD, master)

* Sat Dec 9 2017 Tommy Reynolds
- (9f6af0d) History (tag: v2.2.25, origin/master, origin/HEAD)

* Sat Dec 9 2017 Tommy Reynolds
- (d1deba5) OK, I couldn't stand quoting qnything that didn't need quoting.

* Fri Dec 8 2017 Tommy Reynolds
- (6fbdf31) Modernize (more or less) the markdown utility RPM name.

* Fri Dec 8 2017 Tommy Reynolds
- (c311ba9) Merge branch 'master' of github.com:megacoder/vlm-tool

* Fri Dec 8 2017 Tommy Reynolds
- (65bc0da) Flag avc denials (tag: v2.2.24)

* Fri Dec 8 2017 Tommy Reynolds
- (879d018) Flag "avc: denied" entries

* Fri Dec 8 2017 Tommy Reynolds
- (238a7b8) Flag "drop_caches:" (tag: v2.2.23)

* Fri Dec 8 2017 Tommy Reynolds
- (85ad276) Add "drop_caches:" rule, since seeing that is usually a sign someone is frobbing around when they shouldn't be.

* Thu Nov 30 2017 Tommy Reynolds
- (636f70f) Merge branch 'master' of github.com:megacoder/vlm-tool

* Thu Nov 30 2017 Tommy Reynolds
- (d7157b9) History (tag: v2.2.22)

* Thu Nov 30 2017 Tommy Reynolds
- (20c2606) Quote everything except the program name.

* Wed Nov 29 2017 Tommy Reynolds
- (6ae7c15) Merge branch 'master' of github.com:megacoder/vlm-tool

* Wed Nov 29 2017 Tommy Reynolds
- (dd94a92) Added command line reconstruction if marking is enabled.

* Wed Nov 29 2017 Tommy Reynolds
- (4f5bd18) Reconstruct command line if marking is enabled. (tag: v2.2.21)

* Tue Jun 13 2017 Tommy Reynolds
- (5879249) Detritus

* Sat Jun 3 2017 Tommy Reynolds
- (ec05288) History (tag: v2.2.20)

* Sat Jun 3 2017 Tommy Reynolds
- (33b28d0) Added "ENERGY_PERF_BIAS:" to spot potential CPU throttling performance issues.

* Sun May 21 2017 Tommy Reynolds
- (7e38aa1) Recognize "eth0:1" too

* Sun May 21 2017 Tommy Reynolds
- (c95e029) Updated

* Sun May 21 2017 Tommy Reynolds
- (6988600) Bumped version. (tag: v2.2.19)

* Sun May 21 2017 Tommy Reynolds
- (e8284ed) Notice Ethernet references.

* Sun May 21 2017 Tommy Reynolds
- (3ab102d) Notice Ethernet references.

* Sun May 21 2017 Tommy Reynolds
- (3fd9079) Highlith "o2hbmonitor" if present

* Thu May 11 2017 Tommy Reynolds
- (ef1a1a5) Update ChangeLog (tag: 2.2.18)

* Thu May 11 2017 Tommy Reynolds
- (cc82f0e) Bumped version number

* Thu May 11 2017 Tommy Reynolds
- (d57fe75) Added "certificate(s)" to help detect SSL validation issues.

* Fri Apr 28 2017 Tommy Reynolds
- (12b560d) Bumped (tag: 2.2.17)

* Fri Apr 28 2017 Tommy Reynolds
- (37dc7d9) Must use zcat for in-place expanding.

* Fri Apr 28 2017 Tommy Reynolds
- (5c3a725) uncompress(1) does not honor '--'

* Fri Apr 28 2017 Tommy Reynolds
- (f27ab8e) Bumped (tag: 2.2.16)

* Fri Apr 28 2017 Tommy Reynolds
- (2e5f434) Bumped version number

* Fri Apr 28 2017 Tommy Reynolds
- (06f42db) Added ".Z" uncompress support.

* Sun Dec 4 2016 Tommy Reynolds
- (d97a9c4) 2.2.15 (tag: 2.2.15)

* Sun Dec 4 2016 Tommy Reynolds
- (15dd739) Flag NFS server OK/not responding flags.

* Sat Sep 3 2016 Tommy Reynolds
- (3a7b375) Try to flag software raid (md:,md<n>,mds*) lines (tag: 2.2.14)

* Thu Jul 21 2016 Tommy Reynolds
- (0dbcedc) Merge branch 'devel'

* Thu Jul 21 2016 Tommy Reynolds
- (eedce9a) Try to flag D-state processes.

* Thu Jul 14 2016 Tommy Reynolds
- (7a95a66) Added "badness" as a search term. (tag: 2.2.13)

* Wed Jun 29 2016 Tommy Reynolds
- (3a7c8bc) Added keywords like "dropped", "mismatch" and added optional unit number to "scsi" recognition.

* Fri Jun 24 2016 Tommy Reynolds
- (4fce848) Try to identify D-state processes.  They have a "D" as the third token of a "kernel:" line.

* Sat Apr 9 2016 Tommy Reynolds
- (76e010d) Change log. (tag: 2.2.12)

* Sat Apr 9 2016 Tommy Reynolds
- (8235e56) Don't grumble about posix_fadvise(2) failing.

* Wed Feb 17 2016 Tommy Reynolds
- (8ac1040) Updated changelog

* Wed Feb 17 2016 Tommy Reynolds
- (2bab99b) Don't issue fadvise() for piped input, used if log file is compressed.

* Wed Oct 14 2015 Tommy Reynolds
- (585d97a) Added "unauthorized" as a red flag (tag: 2.2.11, tag: 2.2.10)

* Wed Oct 14 2015 Tommy Reynolds
- (056cf30) Bumped version

* Wed Oct 14 2015 Tommy Reynolds
- (b0465a0) Versioning. (tag: 2.2.9)

* Wed Oct 14 2015 Tommy Reynolds
- (58256a5) Highlight D-bus references

* Tue Sep 22 2015 Tommy Reynolds
- (8e346e7) Bumped revision (tag: 2.2.8)

* Tue Sep 22 2015 Tommy Reynolds
- (51b6a8a) Allocate storage pool in larger chunks.

* Tue Sep 22 2015 Tommy Reynolds
- (c71fb25) Noted use of fadvise (tag: 2.2.7)

* Tue Sep 22 2015 Tommy Reynolds
- (e4b2270) Bumped version number

* Tue Sep 22 2015 Tommy Reynolds
- (5c3f154) Use fadvise trying to preload as much of the messages file as we can into the page cache to improve performance.

* Fri Jul 10 2015 Tommy Reynolds
- (9764c14) bump (tag: 2.2.6)

* Fri Jul 10 2015 Tommy Reynolds
- (067c3fc) Highlight 'leap second' notices.

* Fri Jul 10 2015 Tommy Reynolds
- (814c756) Try to be not fooled by an enbedded 'hang' or 'hung' string.

* Sat Jun 20 2015 Tommy Reynolds
- (1367d1f) Bumped revision number.

* Sun Apr 26 2015 Tommy Reynolds
- (127e8f1) Corrected man(1) page: the '-d' switch actually takes '-d <region>' and defaults to USA.

* Sun Mar 1 2015 Tommy Reynolds
- (0ef7e7c) Indent ignored patterns by a tab, not a space.

* Wed Feb 25 2015 Tommy Reynolds
- (8b2c4fc) Updated changelog. (tag: 2.2.4)

* Wed Feb 25 2015 Tommy Reynolds
- (17b202f) Bumped rev

* Wed Feb 25 2015 Tommy Reynolds
- (4f4699d) Show filename and line number when complaining about timestamp format.

* Wed Feb 25 2015 Tommy Reynolds
- (cd2d868) Try using "markdown2" else "markdown" else punt.

* Wed Feb 25 2015 Tommy Reynolds
- (adba464) Softened required packages for build.

* Sat Jan 17 2015 Tommy Reynolds
- (8d3ebdd) Updated change log. (tag: 2.2.3)

* Sat Jan 17 2015 Tommy Reynolds
- (9db4318) Bumped revision.

* Sat Jan 17 2015 Tommy Reynolds
- (fa90d43) Mark kernel version information.  It's not an error, but is useful in all contexts.


vlm-tool.py
#!/usr/bin/python
# vim: noet sw=4 ts=4

VERSION	= '@VERSION@'

import	os
import	sys
import	re
import	datetime
import	gzip
import	bz2

class	AnsiColors( object ):

	RESET_ALL      = '\033[0m';
	FOREGROUND_RED = '\033[1;31m';
	FOREGROUND_GREEN = '\033[1;32m';
	FOREGROUND_YELLOW = '\033[1;33m';
	REVERSE_VIDEO  = '\033[7m';

	def	__init__( self ):
		return

	def	reset( self ):
		return AnsiColors.RESET_ALL

	def	emphasis( self ):
		return AnsiColors.FOREGROUND_RED

	def	normal( self ):
		return AnsiColors.FOREGROUND_GREEN

	def	timestamp( self ):
		return AnsiColors.FOREGROUND_YELLOW

	def	reverse( self ):
		return AnsiColors.REVERSE_VIDEO

	def	host_color( self, id ):
		sgr = '\033[1;%dm' % (31 + (id % 6))
		return sgr

class	VlmTool( object ):

	FILTERS =	[
		# Try to keep these in alphabetical order, at least
		r'abnormal',
		r'abo?rt',
		r'access beyond',
		r'asked for',
		r'bcm5709',
		r'blocked for more',
		r'becomes',
		r'bond',
		r'call trace:',
		r'clocksource',
		r'conflict',
		r'crs',
		r'css',
		r'device-mapper',
		r'(de)?bug',
		r'dlm',
		r'disagrees.*symbol.*',
		r'don\'t',
		r'end_request',
		r'error',
		r'fail',
		r'fenc(e,ing)',
		r'fsck',
		r'\bh[au]ng',
		r'illegal:',
		r'init:',
		r'invalid:',
		r'i?scsi',
		r'kdump:?',
		r'kill',
		r'link',
		r'linux version',
		r'lpfc',
		r'multipathd',
		r'netdev',
		r'nfs:.*not responding',
		r'nfsd:',
		r'no space left',
		r'not supported',
		r'\bnmi',
		r'o2',
		r'ocfs',
		r'only',
		r'oom',
		r'oops',
		r'out.*of.*memory',
		r'own address',
		r'panic',
		r'please',
		r'page allocation failure',
		r'power states',
		r'qla.*loop',
		r'rds/(ib)?',
		r'read-?only',
		r'register(ed)?',
		r'require',
		r'restart',
		r'segfault',
		r'selinux',
		r'signal',
		r'sysrq.*:',
		r'taint',
		r'terminated',
		r'time-?out',
		r'unable',
		r'unknown symbol.*',
		r'unsupported',
		r'vfs:',
		r'warning',
		r'watchdog',
	]

	def __init__( self, dont_populate = True, mark = False, colorize = False ):
		self.colorize = colorize
		self.accepted = 0
		self.rejected = 0
		self.ignored  = 0
		self.entries  = []
		self.filters  = []
		self.ruleno   = 0		# State for incremental searches
		self.maxrules = 0
		self.mark     = mark
		if not dont_populate:
			self.add_filter_set( VlmTool.FILTERS )
		return

	def add_filter( self, pattern ):
		if self.colorize:
			pattern = r'^(.*)(%s)(.*)$' % pattern
			# print >>sys.stderr, 'pattern %s' % pattern
		mo = re.compile( pattern, re.IGNORECASE )
		self.filters.append( mo )
		self.maxrules += 1
		return	mo

	def show_rule( self, mo ):
		if mo is None:
			return ''
		return mo.re.pattern

	def add_filter_set( self, filters ):
		for filter in filters:
			self.add_filter( filter )
		return

	def filter_matches( self, text, ruleno ):
		matchObject = self.filters[ruleno].search( text )
		if matchObject is not None:
			return matchObject
		return None

	def apply_filters( self, line ):
		for i in xrange( 0, self.maxrules ):
			mo = self.filter_matches( line, self.ruleno )
			if mo is not None:
				return mo
			self.ruleno = (self.ruleno+1) % self.maxrules
		return None

	def	get_parts( self, mo ):
		if mo is None:
			l = '???'
			m = '???'
			r = '???'
		else:
			l = mo.string[mo.start(1):mo.end(1)]
			m = mo.string[mo.start(2):mo.end(2)]
			r = mo.string[mo.start(3):mo.end(3)]
		return (l, m, r)

	def date_to_bin( self, ts ):
		try:
			retval = datetime.datetime.strptime(
				ts,
				'%b %d %H:%M:%S'
			)
		except:
			raise ValueError
		return retval

	def ingest_from( self, fyle ):
		for line in fyle:
			line = line.rstrip()
			if len(line) == 0: continue
			timestamp = line[0:15]
			try:
				ts = self.date_to_bin( timestamp )
			except:
				# Silently ignore badly-formatted line
				self.ignored += 1
				continue
			for i in xrange( 17, len(line) ):
				if line[i].isspace():
					host = line[16:i]
					resid = line[i:]
					break
			mo = self.apply_filters( resid )
			if mo is not None:
				# Rule hits are always accepted
				self.entries.append( (ts, mo, timestamp, host, resid) )
				self.accepted += 1
			else:
				if self.mark:
					# Non-hits are only kept if we are marking or colorizing
					self.entries.append( (ts, mo, timestamp, host, resid) )
				self.rejected += 1
		return

	def ingest( self, fn ):
		if fn.endswith( '.bz2' ):
			openwith = bz2.open
		elif fn.endswith( '.bz' ):
			openwith = gzip.open
		else:
			openwith = open
		try:
			f = openwith( fn, 'rt' )
		except Exception, e:
			print >>sys.stderr, "Cannot open '%s' for injesting." % fn
			raise e
		self.ingest_from( f )
		f.close()
		return

	def dump_stats( self, out = sys.stdout ):
		total = self.accepted + self.rejected + self.ignored
		l = len( '%u' % total )
		u_fmt = '%' + str(l) + 'u %s'
		print >>out, '%s' % '-' * l
		print >>out, u_fmt % ( self.accepted, 'entries accepted' )
		print >>out, u_fmt % ( self.ignored,  'entries ignored' )
		print >>out, u_fmt % ( self.rejected, 'entries rejected' )
		print >>out, '%s' % '=' * l
		print >>out, u_fmt % ( total, 'all entries' ),
		print >>out, '(%.f%% flagged)' % ((self.accepted * 100.0) / total)
		return

	def date_to_bin( self, ts ):
		return datetime.datetime.strptime(
			ts,
			'%b %d %H:%M:%S'
		)

	def sort( self ):
		self.entries.sort( key = lambda (ts,mo,timestamp,host,resid): ts )
		return

	def every( self ):
		for (t,mo,timestamp,host,resid) in self.entries:
			yield (t,mo,timestamp,host,resid)
		return

	def is_marked( self, mo ):
		if mo is None:
			retval = False
		else:
			retval = True
		return retval

	def	list_filter_set( self, out = sys.stdout ):
		for filter in self.filters:
			print >>out, "%s" % filter.pattern
		return

	def	postprocess( self ):
		starters = []
		starters.append(self.add_filter( r'kernel:.*unable to handle kernel' ))
		starters.append(self.add_filter( r'call trace:' ))
		starters.append(self.add_filter( r'kernel: eeek!' ))
		body  = self.add_filter( r'kernel:' )
		i = 0
		n = len( self.entries )
		while i < n:
			(ts, mo, timestamp, host, resid) = self.entries[i]
			for starter in starters:
				mo = starter.search( resid )
				if mo is not None:
					break
			if mo is None:
				i += 1
				continue
			self.entries[ i ] = (ts,mo,timestamp,host,resid)
			# Begin a matching clause
			i += 1
			while i < n:
				(ts, mo, timestamp,host,resid) = self.entries[i]
				mo = body.search( resid )
				if mo is None:
					break
				self.entries[ i ] = (ts,mo,timestamp,host,resid)
				i += 1
		return

if __name__ == '__main__':
	import	optparse
	prog = os.path.basename( sys.argv[0] )
	p = optparse.OptionParser(
		prog = prog,
		version = '%%prog v%s' % VERSION,
		usage = 'usage: %prog [options] [messages..]',
		description = """Read and filter groups of /var/log/messages files.""",
		epilog="""Consult the man(1) page for more information."""
	)
	p.add_option(
		'-a',
		'--add',
		dest='filters',
		metavar='re',
		default = [],
		action='append',
		help = 'Python regular expression'
	)
	p.add_option(
		'-b',
		'--bulk',
		dest='from_here',
		metavar='FILE',
		default = None,
		help = 'Load bulk filters from here, one per line.'
	)
	p.add_option(
		'-c',
		'--colorize',
		dest='colorize',
		default = False,
		action='store_true',
		help = 'Use colors (implies -m).'
	)
	p.add_option(
		'-l',
		'--list',
		dest='list_rules',
		default = False,
		action = 'store_true',
		help = 'Write rule set to stdout.'
	)
	p.add_option(
		'-m',
		'--mark',
		dest='mark',
		default = False,
		action = 'store_true',
		help = 'Output all lines, marking matches.'
	)
	p.add_option(
		'-n',
		'--nofilters',
		dest='no_filters',
		default = False,
		action = 'store_true',
		help = 'Do not populate standard filters.'
	)
	p.add_option(
		'-o',
		'--out',
		dest='ofile',
		metavar='FILE',
		default = None,
		help = 'Write output here, else to stdout.'
	)
	p.add_option(
		'-r',
		'--rules',
		dest='show_rule',
		default = False,
		action = 'store_true',
		help = 'Show matching rule for line (implies -m).'
	)
	p.add_option(
		'-s',
		'--stats',
		dest='show_stats',
		default = False,
		action = 'store_true',
		help = 'Show counts of filter hits.'
	)
	def_thumb = '-'
	p.add_option(
		'-t',
		'--thumb',
		dest='thumb',
		default = def_thumb,
		metavar = 'THUMB',
		help = 'Mark matching lines with this; default="%s".' % def_thumb
	)
	(opts,args) = p.parse_args()
	if opts.ofile is not None:
		out = open( opts.ofile, 'wt' )
	else:
		out = sys.stdout
	opts.mark |= opts.colorize
	vt = VlmTool(
		dont_populate = opts.no_filters,
		mark = opts.mark,
		colorize = opts.colorize
	)
	vt.add_filter_set( opts.filters )
	if( opts.from_here is not None ):
		try:
			f = open( opts.from_here, 'rt' )
		except Exception, e:
			print >>sys.stderr, "Could not read '%s'; bye." % opts.from_here
		bulk = []
		for line in f:
			bulk.append( line.strip() )
		f.close()
		vt.add_filter_set( bulk )
	if( opts.list_rules ):
		vt.list_filter_set()
		sys.exit(0)
	if len(args) == 0:
		for f in os.listdir( '/var/log' ):
			if f.startswith( 'messages' ):
				vt.ingest( os.path.join( '/var/log', f ) )
	else:
		for f in args:
			if f == '-':
				vt.ingest_from( sys.stdin )
			else:
				vt.ingest( f )
	vt.sort()
	thumb = opts.thumb
	nothumb = ' ' * len( thumb )
	if opts.colorize:
		opts.mark = True
		ac = AnsiColors()
	vt.postprocess()
	hosts = {}
	host_number = 1
	for (ts,mo,timestamp,host,resid) in vt.every():
		if opts.mark:
			# Will get both marked and unmarked lines here
			marked = vt.is_marked( mo )
			if marked is False:
				leadin = nothumb
			else:
				leadin = thumb
			print >>out, '%s ' % leadin,
			if opts.show_rule:
				if not marked:
					rule = ''
				else:
					rule = vt.show_rule( mo )
				print >>out, '%-15.15s ' % rule,
			if opts.colorize:
				try:
					color = hosts[ host ]
				except KeyError, e:
					color = host_number
					hosts[ host ] = host_number
					host_number += 1
				host = '%s%s%s' %	(
					ac.host_color(color),
					host,
					ac.reset()
				)
			if marked and opts.colorize:
				l,m,r = vt.get_parts( mo )
				resid = '%s%s%s%s%s' % (
					l,
					ac.emphasis(),
					m,
					ac.reset(),
					r
				)
		print >>out, '%s %s %s' % (timestamp,host,resid)
	if opts.show_stats:
		vt.dump_stats( out )
	sys.exit(0)
